homeassistant/components/sensor/imap_email_content.py

```diff
@@ -0,0 +1,249 @@
+
+    def read_next(self):
+        """Read the next email from the email server."""
+        import imaplib
+        try:
+            self.connection.select()
+
+            if len(self._unread_ids) == 0:
+                search = "SINCE {0:%d-%b-%y}".format(datetime.date.today())
+                if self._last_id is not None:
+                    search = "UID {}:*".format(self._last_id)
+
+                _, data = self.connection.uid("search", None, search)
+                self._unread_ids = deque(data[0].split())
+
+            while len(self._unread_ids) > 0:
+                message_uid = self._unread_ids.popleft()
+                if self._last_id is None or int(message_uid) > self._last_id:
+                    self._last_id = int(message_uid)
+                    return self._fetch_message(message_uid)
+
+        except imaplib.IMAP4.error:
+            _LOGGER.info(
+                "Connection to %s lost, attempting to reconnect",
+                self._server)
+            try:
+                self.connect()
+            except imaplib.IMAP4.error:
+                _LOGGER.error("Failed to reconnect.")
+
+
+class EmailContentSensor(Entity):
+    """Representation of an EMail sensor."""
+
+    # pylint: disable=too-many-arguments
+    # pylint: disable=too-many-instance-attributes
+    def __init__(self,
+                 hass,
+                 email_reader,
+                 name,
+                 allowed_senders,
+                 value_template):
+        """Initialize the sensor."""
+        self.hass = hass
+        self._email_reader = email_reader
+        self._name = name
+        self._allowed_senders = allowed_senders
+        self._value_template = value_template
+        self._last_id = None
+        self._message = None
+        self._state_attributes = None
+        self.connected = self._email_reader.connect()
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return self._name
+
+    @property
+    def state(self):
+        """Return the number of unread emails."""
+        return self._message
+
+    @property
+    def state_attributes(self):
+        """Return other state attributes for the message."""
+        return self._state_attributes
+
+    def render_template(self, email_message):
+        """Render the message template."""
+        variables = {
+            ATTR_FROM: EmailContentSensor.get_msg_sender(email_message),
+            ATTR_SUBJECT: EmailContentSensor.get_msg_subject(email_message),
+            ATTR_DATE: email_message['Date'],
+            ATTR_BODY: EmailContentSensor.get_msg_text(email_message)
+        }
+        return template.render(self.hass, self._value_template, variables)
+
+    def sender_allowed(self, email_message):
+        """Check if the sender is in the allowed senders list."""
+        return EmailContentSensor.get_msg_sender(email_message).upper() in (
+            sender.upper() for sender in self._allowed_senders)
```

Comment:
If we always compare allowed_senders uppercase, why not store it as uppercase in `__init__`?